compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/hotspot/aarch64/shenandoah/AArch64HotSpotShenandoahLoadRefBarrierOp.java

```diff
@@ -0,0 +1,233 @@
+import jdk.graal.compiler.hotspot.replacements.HotSpotReplacementsUtil;
+import jdk.graal.compiler.lir.LIRInstructionClass;
+import jdk.graal.compiler.lir.aarch64.AArch64AddressValue;
+import jdk.graal.compiler.lir.aarch64.AArch64Call;
+import jdk.graal.compiler.lir.aarch64.AArch64LIRInstruction;
+import jdk.graal.compiler.lir.asm.CompilationResultBuilder;
+import jdk.graal.compiler.nodes.gc.shenandoah.ShenandoahLoadRefBarrierNode;
+
+import jdk.vm.ci.code.CallingConvention;
+import jdk.vm.ci.code.Register;
+import jdk.vm.ci.meta.AllocatableValue;
+
+import static jdk.graal.compiler.lir.LIRInstruction.OperandFlag.COMPOSITE;
+import static jdk.graal.compiler.lir.LIRInstruction.OperandFlag.REG;
+import static jdk.vm.ci.code.ValueUtil.asRegister;
+
+/**
+ * AArch64 backend for the Shenandoah load-reference barrier.
+ */
+public class AArch64HotSpotShenandoahLoadRefBarrierOp extends AArch64LIRInstruction {
+    public static final LIRInstructionClass<AArch64HotSpotShenandoahLoadRefBarrierOp> TYPE = LIRInstructionClass.create(AArch64HotSpotShenandoahLoadRefBarrierOp.class);
+
+    enum GCStateBitPos {
+        // Heap has forwarded objects: needs LRB barriers.
+        HAS_FORWARDED_BITPOS(0),
+
+        // Heap is under marking: needs SATB barriers.
+        // For generational mode, it means either young or old marking, or both.
+        MARKING_BITPOS(1),
+
+        // Heap is under evacuation: needs LRB barriers. (Set together with HAS_FORWARDED)
+        EVACUATION_BITPOS(2),
+
+        // Heap is under updating: needs no additional barriers.
+        UPDATE_REFS_BITPOS(3),
+
+        // Heap is under weak-reference/roots processing: needs weak-LRB barriers.
+        WEAK_ROOTS_BITPOS(4),
+
+        // Young regions are under marking, need SATB barriers.
+        YOUNG_MARKING_BITPOS(5),
+
+        // Old regions are under marking, need SATB barriers.
+        OLD_MARKING_BITPOS(6);
+
+        private final int value;
+
+        GCStateBitPos(int val) {
+            this.value = val;
+        }
+
+        public int getValue() {
+            return this.value;
+        }
+    }
+
+    enum GCState {
+        HAS_FORWARDED(1 << GCStateBitPos.HAS_FORWARDED_BITPOS.value),
+        MARKING(1 << GCStateBitPos.MARKING_BITPOS.value),
+        EVACUATION(1 << GCStateBitPos.EVACUATION_BITPOS.value),
+        UPDATE_REFS(1 << GCStateBitPos.UPDATE_REFS_BITPOS.value),
+        WEAK_ROOTS(1 << GCStateBitPos.WEAK_ROOTS_BITPOS.value),
+        YOUNG_MARKING(1 << GCStateBitPos.YOUNG_MARKING_BITPOS.value),
+        OLD_MARKING(1 << GCStateBitPos.OLD_MARKING_BITPOS.value);
+
+        private final int value;
+
+        GCState(int val) {
+            this.value = val;
+        }
+
+        public int getValue() {
+            return this.value;
+        }
+    }
+
+    private final HotSpotProviders providers;
+    private final GraalHotSpotVMConfig config;
+
+    @Def({REG}) protected AllocatableValue result;
+    @Use({REG}) protected AllocatableValue object;
```

Comment:
Note that here `object` and `result` may be allocated to the same register, which seems fine

Reply:
> Note that here `object` and `result` may be allocated to the same register, which seems fine

Yes, that is intentional. Thanks.